### PR TITLE
Fun with lambdas etc

### PR DESCRIPTION
**Pull request by vagababov**
- some code optimization for the lease processing code
- tests for the 'complex function'
- other nits

/assign mattmoor

### CODE REVIEW

File: test/ha/ha.go [hunk 1]
```diff
@@ -26,34 +26,48 @@ import (
...
+func extractDeployment(pod string) string {
+	if x := strings.LastIndexFunc(pod, countingRFind('-', 2)); x != -1 {
+		return pod[:x]
+	}
+	return ""
+}
+
 // GetLeaders collects all of the leader pods from the specified deployment.
-func GetLeaders(t *testing.T, client *test.KubeClient, deploymentName, namespace string) ([]string, error) {
+func GetLeaders(t *testing.T, client *test.KubeClient, deploymentName, namespace string) (sets.String, error) {
```
Comment (mattmoor) [inline test/ha/ha.go:52]: This is a breaking change.  `len(leaders)` on line 88 may never be satisfied if this deduplicates

Comment (vagababov) [reply test/ha/ha.go]: Done

File: test/ha/ha.go [hunk 2]
```diff
@@ -26,34 +26,49 @@ import (
...
+
 // GetLeaders collects all of the leader pods from the specified deployment.
+// GetLeaders will return duplicate pods by design.
 func GetLeaders(t *testing.T, client *test.KubeClient, deploymentName, namespace string) ([]string, error) {
 	leases, err := client.Kube.CoordinationV1().Leases(namespace).List(metav1.ListOptions{})
 	if err != nil {
 		return nil, fmt.Errorf("error getting leases for deployment %q: %w", deploymentName, err)
 	}
-	var pods []string
+	ret := make([]string, 0, len(leases.Items))
```
Comment (mattmoor) [inline test/ha/ha.go:58]: Which is the greater evil: overallocation or resizing? 🤔

Comment (vagababov) [reply test/ha/ha.go:58]: resizing, since resize will probably overallocate as well! (newSize *=2 most of the time).

**Pull request merged**